### ISSUE TITLE
Compatibility with Boost motor?

### ISSUE THREAD

**Issue by nathankellenicki** [JorgePe/Puppy#1]
Hi Jorge, thanks for your work on this app! Great work. I was wondering, have you managed to get the Boost motor working with it yet? It seems when working on the Powered Up hub, the normal commands don't work with it in my experiments.

Thanks,
Nathan

**Comment by JorgePe** [comment]
Hi Nathan.
No time for it yet... my life has been crazy last months,  had just a few
hours at lunch to start with Powered Up and never had the time to finish
BOOST.

I hope the commands work with just slight modifications (like port numbers)
since someone already made the color sensor work as well.

**Comment by nathankellenicki** [comment]
Hi Jorge, no problem. :) 

Yes indeed, I've managed to get the color sensor working, along with all previous WeDo sensors. I've also figured out the commands to start receiving button press notifications on the Powered Up remote. But the Boost motor seems different - usual sensor activation methods don't seem to work, and neither do the normal motor commands.

I'll keep messing around anyway!

Thanks

**Comment by JorgePe** [comment]
You have the original firmware, I suppose?
I know that there is already at least one firmware update but no idea what
it changes.
I believe that this new smart hub ("HUB NO.4") is not an evolution from the
BOOST "Move Hub". I think they were developed more or less at the same time
but BOOST received more attention because it was chosen to be released to
market first. If I'm right, we'll have to wait for LEGO developers to
release an equivalent firmware.

**Comment by JorgePe** [comment]
Hi.
Took the BOOST motor with me today. It is recognized, only difference is
the port (00 = A and 01 = B but you already know that).
Also no luck.
I believe it is not implemented in firmware. Not sure why beecause as you
found out the Color/Dist sensor is already implemented and I'm not aware of
any current know set that uses it.
Have to wait for a future hardware.

**Comment by nathankellenicki** [comment]
Ah unfortunate.

Incidentally, if you connect the Boost motor to the Powered Up hub, and link the Powered Up remote to it, it *can* drive the motor in an on/off capacity, the same as the WeDo 2.0 motor. So there seems to be some kind of support there?

**Comment by JorgePe** [comment]
You are a lucky guy... my wife gave me my advanced birthday gift, a train.
So I now have a controller... and I just arrive from the near by market
with 4x AAA batteries.
You are absolutely right.
But as far as I know the train motors are controlled as the WeDo 2 motors
and I tried that same commands from my linux laptop without success.

Damn, I'll never have a good night sleep again this year :)

**Comment by nathankellenicki** [comment]
Haha, I apologise for that! :)

Well, I've noticed that actually the train motors *aren't* identical to the WeDo 2.0 motors. The WeDo 2.0 motor is identified with peripheral id 0x01, but the train motor is peripheral id 0x02, and the boost motor is peripheral id 0x26. This is sent in an 0x45 message when the motor is plugged in.

Also you notice that the remote control has different behaviour depending on what motor is plugged into the PUP hub. If it's a train motor, pressing up on the remote will increase the speed. But if its a WeDo or Boost motor, the motor will only be active for as long as the up button is pressed (releasing will stop the motor).

I notice from your projects you have a BLE sniffer, are you capable of sniffing the commands the remote sends to the PUP hub at some point?

**Comment by nathankellenicki** [comment]
See here for the peripheral id list I've put together: https://github.com/nathankunicki/node-lpf2/blob/master/consts.ts#L9-L21

**Comment by JorgePe** [comment]
I am. At night, at home, where my bluetooth environment is less poluted and
no kids or work mates are interrupting... that's why I'll never sleep again
:D

**Comment by JorgePe** [comment]
Just an extra cent:
the wedo 2 and the train motors are different in hardware level, they even
have different AutoID (already got that hacked).
But I can control both with the same command:

0800810011510060

payload size = 08 |  00 or 01 or probably anything  | port command = 81 |
port/connector ID = 00 |  command metadata = 11 | command = 51 |  command
args = 00 60 where 60 = Duty Cycle or speed


but not the BOOST motor because it is a different type of motor (device
type 26)

So the powered Up train command, after making the BLE connection, probably
makes something like put the output port in a new mode that treats motors
different than 02 (train motors) as generic motors.

**Comment by nathankellenicki** [comment]
Yep agreed (I think we're saying same thing but using different terms - auto id/peripheral id) :)

I'm sending pretty much the same command: https://github.com/nathankunicki/node-lpf2/blob/master/lpf2hub.ts#L127

I tried enumerating port modes 0-10 and activating them, but it seemed to have no effect sadly

**Comment by JorgePe** [comment]
My father in law birthday is today but I'll try to sniff the command
conenction tomorrow, or at least this week.
We will find the mode. But we will only get the ON/OFF option... unless we
are lucky to understand other possible modes, we don't get the angle
control.

**Comment by nathankellenicki** [comment]
That makes sense. Hopefully we can deduce the modes for speed/angle control. Thanks for your help. :)

**Comment by JorgePe** [comment]
Run just a few captures... not much new data.

The Train Command connects as a slave to the Powered Up hub. It has a Texas
Instruments BT address.

After it connects there are some queries to identify port 0, then this
command is issued:
0a004100000100000001
with a reply from the hub:
0a004700000100000001

then the same is repeated to the port 1
I never saw any activity when pressing the command buttons - I'm not sure
that the speed controls use BT BLE. But need much more captures to be sure.

I tried the same command from my linux laptop and got the same answer:

[90:84:2B:06:AB:5D][LE]> char-write-req 0x0f 0100
Characteristic value was written successfully
Notification handle = 0x000e value: 0f 00 04 00 01 26 00 00 00 00 10 00 00
00 10
Notification handle = 0x000e value: 0f 00 04 32 01 17 00 00 00 00 10 00 00
00 10
Notification handle = 0x000e value: 0f 00 04 3b 01 15 00 02 00 00 00 02 00
00 00
Notification handle = 0x000e value: 0f 00 04 3c 01 14 00 02 00 00 00 02 00
00 00
[90:84:2B:06:AB:5D][LE]> char-write-req 0x0e 0a004100000100000001
Characteristic value was written successfully
Notification handle = 0x000e value: 0a 00 47 00 00 01 00 00 00 01

sending a motor command is accepted but does nothing:

[90:84:2B:06:AB:5D][LE]> char-write-req 0x0e 0800810011510064
Characteristic value was written successfully
Notification handle = 0x000e value: 05 00 82 00 0a


If I try the Tacho motor command from BOOST it returns an error:

[90:84:2B:06:AB:5D][LE]> char-write-req 0x0e 0C0081001109000164647F03
Characteristic value was written successfully
Notification handle = 0x000e value: 05 00 82 00 0a
Notification handle = 0x000e value: 05 00 05 81 06


I wonder why the WeDo2 motor command is accepted.

I have a few other messages but captures are not consistent, need to try a
few more.

**Comment by nathankellenicki** [comment]
Ah good work, I find it a bit hard to follow, but makes sense! Just to be clear when you do it from your laptop you are connecting to the remote, not the hub?

For what its worth I already figured out the command buttons on the remote through trial and error. Port 0 is the left set of buttons, port 1 is the set on the right. They are advertised with a normal 0x45 command as being of sensor type 0x37. If you then activate that sensor with mode 0x00 you start getting button press events. See code here:

https://github.com/nathankunicki/node-lpf2/blob/master/hub.ts#L289
https://github.com/nathankunicki/node-lpf2/blob/master/lpf2hub.ts#L366

So I can connect to the remote from my laptop and start getting button press events (I made a remote for Vernie here :) https://github.com/nathankunicki/node-lpf2/blob/master/examples/vernie_remote.js)

But I'm unsure how those button press events activate a boost motor when the remote is connected to a hub! Weird that you don't see any traffic happening when you press the buttons?

**Comment by JorgePe** [comment]
To the hub.

Good find on the 0x37, I was looking for a new notification I captured
after connection:
0f:00:04:00:01:37:00:00:00:00:10:00:00:00:10

This sniffer is tricky, almost black voodoo. I'm going to try a pre-release
of the new code, hope it works better.

So 37 is the ID of the Train Remote and it is associated to port 0 like my
BOOST motor (why the same port?)

I taught the remote, being a slave, wasn't connectable but after reading
your reply I tried from my laptop and got a connection.

it also implements same devices as the hub:
17 (RGB LED) at port 34
14 (Voltage) at port 3B

and a new device 38:
38 (?) at port 3C

I'm a bit lost by now and not sure if you already got that but I got the
Remote LED working from y laptop with the same command for the RGB LED of
BOOST, just changing port to 34:

char-write-req 0x0b 080081341151000n

**Comment by JorgePe** [comment]
If I activate this new 38 device at port 3C

char-write-req 0x0b 0a00413C000100000001

I get frequent values from the port 3C:
Notification handle = 0x000b value: 05 00 45 3c ca 
Notification handle = 0x000b value: 05 00 45 3c c8 
Notification handle = 0x000b value: 05 00 45 3c ca 
Notification handle = 0x000b value: 05 00 45 3c c9 
Notification handle = 0x000b value: 05 00 45 3c c8 
Notification handle = 0x000b value: 05 00 45 3c ca 
Notification handle = 0x000b value: 05 00 45 3c c9 
Notification handle = 0x000b value: 05 00 45 3c ca 

Don't know what it is. Connection strength?

**Comment by nathankellenicki** [comment]
Good find on the LED! That was one of the missing things I had in my library for the remote support. Thanks!

I suspect both the remote and the hub can act in both master and slave mode, depending on who connects to who first during the discovery/connection process. That would explain why we can also connect to the remote from a laptop (as a slave).

I think maybe don't look at the buttons and boost motor as "both occupying port 0", I think think of the hub and remote as two separate devices. Port 0 on the remote is connected to the button "sensor", and port 0 on the hub is connected to the boost motor. This way I view the remote as just another hub type - with the same service, characteristics, command types, ports, etc.

I've gone and bought a cheap sniffer from eBay which should arrive from the next few days, don't know how much luck I'll have with it at first. :)

Maybe a good step would be to try and sniff the commands that are send from the remote to the hub when pressing a button and activating a boost motor? I know you said you couldn't find it, but I think there has to be something there, I don't think these devices are capable of anything but BLE...

**Comment by nathankellenicki** [comment]
> Don't know what it is. Connection strength?

Interesting, I'll have to check this out myself this evening. Does the 4th byte ever change? If it does, is there a secret inbuilt tilt sensor? If not, does moving the hub away change the 5th byte (indicating connection strength as you suggest)?

Bizarre that if it was signal strength they'd implement it like that as well as normal rssi values.

**Comment by nathankellenicki** [comment]
Ah, I think 3C/3B is Voltage and Current/Amps. Looking at Andrey Pokhilko's Boost code, he figured this out for the Boost Move Hub: https://github.com/undera/pylgbst/blob/master/pylgbst/constants.py#L19-L20

**Comment by JorgePe** [comment]
Nice. 3C/3B added to my notes.

I'm afraid the Remote will be hard: it is indeed an hub. Not a "smart hub"
like BOOST and Powered Up boot it is a hub since it can work as a slave
with the Powered Up and as a master with our laptops and still presents a
few devices like the RGB LED.
So the Remote probably just extends the Smart Hub and when we press a
button it doesn't send a Motor command, it just sends a button changed
notification that the Smart Hub understands... if that's true, the Smart
Hub only operates the motor when it receives buttons notifications from a
Dumb Hub... and we will need to find a way to pretend we are a Dumb Hub.

The promised SDK would be so great...!

**Comment by JorgePe** [comment]
Finally got a button press capture. When I'm luck, it captures one or two
events, no more than that. I knew it has problems with capturing traffic
but It also captures much more traffic from the hub than from the remote.

05:00:45:00:01 (left + pressed)
05:00:45:00:00 (left + released)

05:00:45:00:ff (left - pressed)
05:00:45:00:00 (left - released)

Perhaps there is more than that but I don't think so - those are exactly
the same events that we receive when connecting our computer to the Remote.

If I'm right there must be some kind of local service running in the
Powered Up hub that reacts to this events and actuates the motor according
to the type of motor that is connected. And we would need to trick the Hub
to believe we are a Dumb Hub / Remote.

**Comment by JorgePe** [comment]
I attach a capture file
[command-03.zip](https://github.com/JorgePe/Puppy/files/2228461/command-03.zip)

**Comment by nathankellenicki** [comment]
Ahhh, I was afraid of that. :) So sadly we have no examples to go on to see how the boost motor is activated. Either we wait for Lego to release an SDK or a BLE spec, or we magically we figure it out from trial and error.

Keeping my fingers crossed!

**Comment by JorgePe** [comment]
I still have hope that a future firmware unblocks the motor.
And I l didn't sniff the LEGO app yet, perhaps there's something hidden
there... I will try at least a few more nights.

**Comment by JorgePe** [comment]
Closing this.
Thanks for the tips... got a new Remote for the EV3
https://youtu.be/fiuWAwkddAA

**Comment by nathankellenicki** [comment]
Hey @JorgePe, just to let you know, myself and @undera managed to figure out that the command to activate a boost motor from the pup hub is:

`[0x07, 0x00, 0x81, port, 0x11, 0x02, speed]`

I'm still trying to find out the commands for timed and angle modes, as well as how to activate rotation sensor notifications. :)

**Comment by JorgePe** [comment]
Good find.
How did you find it? Sniffer? Pure trying?

**Comment by nathankellenicki** [comment]
I did get a sniffer, but it revealed no more from the remote than you already found. I got it through pure brute forcing:

```js
for (let x = 0; x <= 255; x++) {
    for (let y = 0; y <= 255; y++) {
        const data = Buffer.from([0x08, 0x00, 0x81, portObj.value, 0x11, x, y, this._mapSpeed(speed)]);
        console.log(data);
        this._writeMessage(Consts.BLECharacteristics.BOOST_ALL, data);
        await this.sleep(10);
    }
}
```

;)

**Comment by JorgePe** [comment]
:D

**Comment by JorgePe** [comment]
This is turning interesting. Did you tried with two motors?

Just one BOOST motor at port A (00):
0f 00 04 00 01 26 00 00 00 00 10 00 00 00 10

write 07008100110260 Works
(and also 0800810011026000)

Remove.
Just one BOOST motor at port B (01):
0f 00 04 01 01 26 00 00 00 00 10 00 00 00 10

write 07008101110260 doesn't work
(but it works with 0800810111020060)
Remove

One BOOST motor at port A (00) then another at B (01)

0f 00 04 00 01 26 00 00 00 00 10 00 00 00 10

followed by

**gattool -b 9(gatttool:19058): GLib-WARNING **: 17:36:47.011: Invalid file descriptor.**

and the hub disconnects but never shuts down so I can connect again...
... and then same error and disconnects

it doesn't like both BOOST motors at the same time.

Now if write 08008100110220D0
when I connect 1 motor to A it runs and if i remove and connect to B it runs in opposite turn

and if I connect both they also run for a second before hub shuts down (now it shuts down).

So this command seems to control both motors at same time - just not both connected :)

08 00 81 port 11 02 speedA speedB

Also 'port' usage here is strange, sometimes it seem that doesn't matter if I use 00 or 01

The error happens with the two hubs I have so it doesn't seem battery related but I will try later with fresh batteries on both.

**Comment by JorgePe** [comment]
Also happens with new set of batteries.
And Train Remote control.

**Comment by nathankellenicki** [comment]
Now that is interesting! When we were working out that command we realised that 0x02 matches up with a mode on the Boost hub called CONSTANT_GROUP, usually for driving groups of motors. 0x01 is CONSTANT_SOLO, but attempting to use it results in an error being received. I agree that for driving a single motor 0x02 seems unfit for purpose, and unfortunately I don't have a second Boost motor (yet!) to test.

However, I've brute forced another command, and I'm not sure what to make of it!

`[0x0a, 0x00, 0x81, port, 0x11, 0x60, 0x00, speed, 0x00, 0x00]`

This one appears to work as intended, for both ports (I can't try both at once, but it works with port being either 0x00 or 0x01).

It has a couple of other magical modes though...

`[0x0a, 0x00, 0x81, port, 0x11, 0x60, 0x00, speed, something, 0x00]`

Setting something to...something...appears to run the motor for a specified amount of time before stopping. It doesn't appear to be angle (90, or 0x5a, goes past 90 degrees), and it doesn't appear to be time (It runs for longer than 90ms, and less than 90 seconds!). Any thoughts?

Finally, there's one last mode:

`[0x0a, 0x00, 0x81, port, 0x11, 0x60, 0x00, speed, something, 0x01]`

Setting the last byte to 0x01 (or anything other than 0x00 for that matter) appears to do the same as the above, but then reverse the motor for the same amount of time, then repeat indefinitely. Interesting mode!

**Comment by nathankellenicki** [comment]
And I've also just released that this works for all motor types - Boost, WeDo2, and Train. So it can't be angle, as only the boost motor has a tachometer...

**Comment by JorgePe** [comment]
[0x0a, 0x00, 0x81, port, 0x11, 0x60, 0x00, speed, something, 0x00]

something seems time, but in some kind of ticks instead of ms (one byte instead of two)

**Comment by nathankellenicki** [comment]
Perhaps CPU ticks? I remember reading somewhere that hubs have no concept of "time" - instead the Boost motor, as the only motor with a tachometer, used the number of rotations * speed to estimate how long it had been running for.

Perhaps the PUP Hub has this mode and uses CPU ticks instead? I think I need to do more testing with it

**Comment by JorgePe** [comment]
Perhaps but the hubs have microcontrollers at a know clock so "time" is
possible (as in "timings", not as in  "calendar" because there is no
battery backed up Real Time Clock).
Need more testings but WeDo 1.0 and WeDo 2.0 had timings in multiples of
0.1 seconds.

**Comment by undera** [comment]
Can it just be rotation angle? Boost has the angle rotation mode...

**Comment by nathankellenicki** [comment]
I don't think it's angle - firstly it works on all motors not just the boost motor (the only one with a tachometer), and secondly it doesn't seem to correlate with angle if I put in say 90 or 180

**Comment by nathankellenicki** [comment]
I've also discovered that the last byte is the pause before reversing. :) If you make the last two bytes the same value it'll go for time, pause for time, reverse for time, pause for time, go for time, etc.

**Comment by nathankellenicki** [comment]
I did some sniffing between the Powered Up app and the PUP Hub when in the new Batmobile mode, with 2x WeDo motors attached.

When using the speed sliders, it actually sends the command we discovered a few days ago, the one that you found crashes the boost motors!

[0x08, 0x00, 0x81, 0x39, 0x11, 0x02, 0x20, 0x20]

This sets both motor A and B to speed 0x20.

I notice the port sent is 0x39, I guess this is the port that's used for controlling both motors at once (Similar to the AB port on the Boost Move Hub). I guess this is why your hub was crashing the other day. :)

**Comment by JorgePe** [comment]
0x39 is a virtual port - in BOOST it serves as a combined port. We need to find how to create a virtual port, unless it is created automatically everything two devices with the same AutoID are detected... but that would generate a notification.

Probably the firmware of the Powered Up "HubNO.4" is trying to create a virtual port when it detects two BOOST motors and it crashes

I'm pretty sure it is a firmware bug. Now that the batman car is also selling it's a matter of time until a new firmware arrives.

**Comment by nathankellenicki** [comment]
When you tried with the boost motors did you try with 0x39 as the port? I think above you said you tried with 0x00, 0x01, which might be the problem as you're trying to send a command designed for two devices to a port with only one device?

I suspect 0x39 is static - it gets reported at connection time whether or not there are any devices attached

**Comment by nathankellenicki** [comment]
Ok, have just tried some stuff - that command does not work with Boost motors AT ALL when sent to 0x39.

Boost on A + WeDo on B + 0x39 = Nothing.
WeDo on A + Boost on B + 0x39 = Nothing.
WeDo on A + WeDo on B + 0x39 = Works.

So either it doesn't work with Boost motors on that port, or perhaps there is something to what you were saying about needing two devices with the same AutoID. Can you try with two Boost motors?

**Comment by JorgePe** [comment]
Thanks Nathan, you make me think a bit (I've been distracted with so many things I don't even remind some basic things from last year findings).

No, 0x39 is not static. I was wondering how you guys found it, now I know.

The hub with no devices attached, after enabling notifications (write  0x0100 to 0x0f) , reports only 3 devices:

Notification handle = 0x000e value: 0f 00 04 32 01 17 00 00 00 00 10 00 00 00 10 
Notification handle = 0x000e value: 0f 00 04 3b 01 15 00 02 00 00 00 02 00 00 00 
Notification handle = 0x000e value: 0f 00 04 3c 01 14 00 02 00 00 00 02 00 00 00 


Adding one WeDo 2 motor to A reports another device:

Notification handle = 0x000e value: 0f 00 04 00 01 01 00 00 00 00 00 00 00 00 00

and adding a second WeDo 2 motor to B reports TWO devices:

Notification handle = 0x000e value: 0f 00 04 01 01 01 00 00 00 00 00 00 00 00 00 
Notification handle = 0x000e value: 09 00 04 39 02 01 00 00 01 

Adding a Train motor or a BOOST motor to B instead of a WeDo 2 motor doesn't report '39'

So  the firmware automatically creates the virtual port '39' only when it detects twin motors... and that's when it crashes if using two BOOST motors.

Only have one Train motor... I wonder if it also creates a virtual port with two train motors?

**Comment by JorgePe** [comment]
I cannot try 2 BOOST motors, it disconnects before receiving or accepting anything.

**Comment by nathankellenicki** [comment]
Hah! I understand now, thanks for that insight. :) It seems to be that my hub crashes when I put a boost motor in port B when there is any other type of motor already in port A.

I have three train motors, I'll try dual mode on those when I get home this evening!

**Comment by JorgePe** [comment]
In messages "type 04" that byte after the port ID classifies the type of port:
01 = physical port present
02 = virtual or twin mode port present
00 = none (notfications after device is removed)

**Comment by JorgePe** [comment]
2 Train Motors create a virtual port and doesn't crash:

Notification handle = 0x000e value: 0f 00 04 32 01 17 00 00 00 00 10 00 00 00 10 
Notification handle = 0x000e value: 0f 00 04 3b 01 15 00 02 00 00 00 02 00 00 00 
Notification handle = 0x000e value: 0f 00 04 3c 01 14 00 02 00 00 00 02 00 00 00 
Notification handle = 0x000e value: 0f 00 04 00 01 02 00 00 00 00 00 00 00 00 00 
Notification handle = 0x000e value: 0f 00 04 01 01 02 00 00 00 00 00 00 00 00 00 
Notification handle = 0x000e value: 09 00 04 39 02 02 00 00 01

**Comment by nathankellenicki** [comment]
Nice, your theory was correct! That's very cool. Were you able to move the motors with the 0x02 command?

I wonder if two sensors will create a 0x39 virtual port too (Not sure how that'd even work with sensor notifications...)

**Comment by JorgePe** [comment]
I used the command. Jetro (Hisbabrick Magazine) will be very happy, he has been asking for this feature to control a train with to engines in opposite directions.

**Comment by JorgePe** [comment]
Two BOOST color-dist sensors make it crash :)

Notification handle = 0x000e value: 0f 00 04 32 01 17 00 00 00 00 10 00 00 00 10 
Notification handle = 0x000e value: 0f 00 04 3b 01 15 00 02 00 00 00 02 00 00 00 
Notification handle = 0x000e value: 0f 00 04 3c 01 14 00 02 00 00 00 02 00 00 00 
Notification handle = 0x000e value: 0f 00 04 00 01 25 00 00 00 00 10 00 00 00 10 
(gatttool:2742): GLib-WARNING **: 19:53:50.054: Invalid file descriptor.

**Comment by JorgePe** [comment]
The firmware was only tested with the two motors that are being selled with Powered Up. But they will fix it (although a virtual port with two color sensors is useless, I guess it might be usefull with tilt sensors to add an extra DOF)

**Comment by nathankellenicki** [comment]
Ahh I think it makes sense why it would crash with Boost motors - they are both motors and sensors at the same time. :)

(And as we established, both we and the firmware agree that a virtual port of two sensors makes no sense!)

**Comment by JorgePe** [comment]
The problem is not with the virtual port.
Its with two input devices being found: with two WeDo 2 Sensors (Tilt and Distance) the hub also crashes.

**Comment by nathankellenicki** [comment]
Yep, I was just saying that it makes sense as a Boost motor is also an input device

**Comment by nathankellenicki** [comment]
I discovered today that the 0x60 "swing" command is used by the official Powered Up app when using the Batmobile. There are two buttons, one which turns the Batmobile left to right in a scanning motion, and the second which causes the Batmobile to "retreat" in a swerving motion.

**Comment by JorgePe** [comment]
It's a strange command since we could do that with a combination of others.
I guess they are filling the firmware with commands to reduce the sequence
of commands that an App sends... latency issues?

**Comment by nathankellenicki** [comment]
Possibly...I've been writing a "rampSpeed" function for my lib which, as the name suggests, is intended to steadily ramp the speed from start to finish over a certain amount of time. I've noticed if I do it too fast (ie. 0-100 in about 1 second, so one step every 10ms) it struggles to do it in the time limit. I need to space out the commands to about one every 30-40ms for it to be able to keep up.

Maybe they did it this way to keep the timing precise

**Comment by JorgePe** [comment]
Their decision to make long commands compromises goals like yours. SBrick
has better response times because they have shorter commands... and even
they decided to implement an even shorter option.
To control ramping like you want there must be special commands to the
BOOST motors to use their internal microcontroller, like Dynamixel servos
do.

**Comment by nathankellenicki** [comment]
It is a little annoying but not too bad - I mean if you're ramping from 1-100 in 1 second anyway, you're not going to notice all the individual steps. :) My solution has been to calculate if any of the steps will take less than 50ms and then narrow it down - so my code would do the above example of 1-100 in 20 steps of 5.

I do notice that the WeDo 2.0 Smart Hub has shorter commands. They have no length. For example [setting a motor speed is 5 bytes](https://github.com/nathankellenicki/node-lpf2/blob/master/wedo2smarthub.ts#L126). But they also use different characteristics for everything, whereas Boost/PUP uses one characteristic, so they need bytes for message types too. Maybe they discovered having to use multiple Characteristics had a downside?

**Comment by JorgePe** [comment]
There is at least one reason for using just one characteristic: connection
time is shorter
https://devzone.nordicsemi.com/f/nordic-q-a/12831/one-service-with-multiple-characteristics-or-multiple-services-with-single-characteristic-each

But I also suspect that it uses less resources from the microcontroller
(like less RAM or fewer registers).

On Mon, Aug 6, 2018 at 11:27 AM, Nathan Kellenicki <notifications@github.com

**Comment by JorgePe** [comment]
@nathankellenicki got something new with the BOOST Interactive Motor, but not useful yet.
The Motor uses a serial protocol, 8N1@115200
Immediately after power up it annouces itselft to the BOOST Hub sending
'54 22 00 10 20 B9'
Then it keeps sending 10 bytes containing the encoder (tacho) status:

'D8 00 00 00 00 00 00 00 00 27' or 'D8 00 FF FF FF FF 00 00 00 27'

D8 is a start message signal, 27 is a end message signal and it is the one's complement of D8.
The second byte is 00 when the motor is stopped and reflects directions and I think speed when the motor is rotating.
The 3rd, 4th, 5th and 6th bytes are the encoder position and it is the same value that the BOOST shows when it receives a notification. But not all positions generate a notification, i.e. I can send 5 commands to rotate 1 degree and receive just 3 notifications while seeing a few more status messages in the serial port.

So I think that I can control this motor with a Arduino or a Raspberry Pi, just need two GPIO pins connected to an H-Bridge to control the motor with a PWM signal and read the encoder status with a UART RX pin.

But I tried to emulate the motor and connected a FDTI cable to the BOOST:

RX  -> 6 (TX)
TX -> 5 (RX)
GND  -> 3 (GND)

I connected to port C (01). The BOOST detects a 09 device:

0f 00 04 01 01 09 00 00 00 00 00 00 00 00 00

I think '09' is just a generic UART device but it's still interesting, we can make our own smart devices if we understand it.

Tried sending the AutoID message through the UART  but nothing.
Tried adding resistors (2k2 between 3-GND and 4-VCC and even 2M2 between the AutoID pins and the VCC and GND... nothing.

If you find more info about Device Type 09 please keep me informed.

**Comment by nathankellenicki** [comment]
Interesting! The behaviour certainly seems to be different with the PUP Hub as no battery how I try to send the sensor subscribe events, I get no notifications from the motor.

I'd be curious to know if the motor is sending the notifications and the hub doesn't know what to do with them, or if the motor isn't sending them at all.

I must admit I haven't tried either emulating a hub or device yet, and my electronics stuff is currently in a container crossing the pacific as I'm in the middle of moving to San Francisco! But when I get them back I'll try this stuff. :)

(I kept behind my Lego powered up parts though as I want to keep working on my library!)
